src/share/classes/com/sun/javatest/regtest/exec/Agent.java

```diff
@@ -212,6 +229,59 @@ private Agent(File dir, JDK jdk, List<String> vmOpts, Map<String, String> envVar
         }
     }
 
+    // accept()s a connection over the ServerSocket and then reads handshake bytes
+    // from the connected socket. If the read bytes match the expected handshake bytes
+    // then the connection and handshake is considered successful and the accepted
+    // Socket is returned. If the read bytes don't match the expected handshake bytes
+    // or if the read times out after accept()ing an connection, then this method
+    // closes the accepted connection and returns null.
+    private Socket acceptAndHandshake(final ServerSocket ss, final int handshakeReadTimeout)
+            throws IOException {
+        final Socket s = ss.accept();
+        final byte[] handshakeBytes;
+        int totalRead = 0;
+        try {
+            log("Received connection on port " + ss.getLocalPort() + " from " + s);
+            s.setSoTimeout(handshakeReadTimeout);
+            // read the handshake bytes
+            handshakeBytes = new byte[JTREG_AGENT_HANDSHAKE_MAGIC.length];
+            while (totalRead < JTREG_AGENT_HANDSHAKE_MAGIC.length) {
+                final int numRead = s.getInputStream().read(handshakeBytes,
+                        totalRead, (JTREG_AGENT_HANDSHAKE_MAGIC.length - totalRead));
+                if (numRead == -1) {
+                    break;
+                }
+                totalRead += numRead;
+            }
+        } catch (IOException ioe) {
+            log("closing connection to " + s + " due to: " + ioe);
+            try {
+                s.close();
+            } catch (IOException ignored) {
+                // ignore
+            }
+            return null;
+        }
+        // we don't read more than handshake bytes in this method
+        assert totalRead <= JTREG_AGENT_HANDSHAKE_MAGIC.length : "unexpected number of handshake" +
```

Comment:
It would be an error if totalRead was > JTREG_AGENT_MAGIC.length: according to the logic of the code that reads the magic that cannot happen, ever. Hence the assert.

Reply:
JP, yes understood
I should have been more direct and precise in my comment i.e. 
the assert is not needed — restructure the if statement.

The condition can’t occur and if it did occur then what would that mean?
Is there a problem with either the code generated by Java ? or 
that the java runtime has a problem? or the return from InputStream.read is greater than expected, 
thus the InputStream has a problem?

An AgentServer is (nearly) always launched with “-ea” option — only if jtreg is launched from command line without -ea will it not have asserts enabled.

Looking at the current logic, rather than testing all negative cases, it can be restructured with

if (totalRead != JTREG_AGENT_HANDSHAKE_MAGIC.length) {
   // problem
} else {
   // check for connect message
}

Reply:
So we could do:
```
if (totalRead != JTREG_AGENT_HANDSHAKE_MAGIC.length) {
    assert totalRead < JTREG_AGENT_HANDSHAKE_MAGIC.length
   // problem
} else {
   // check for connect message
}
```

Reply:
I agree with Daniel about the useful of `assert`. I have updated the PR to use Daniel's suggested approach in that snippet.